packages/react-json-tree/src/JSONNestedNode.tsx

```diff
@@ -94,147 +94,116 @@ interface Props extends CircularPropsPassedThroughJSONNestedNode {
   nodeType: string;
   nodeTypeIndicator: string;
   createItemString: (data: any, collectionLimit: number) => string;
-  expandable: boolean;
+  expandable?: boolean;
 }
 
-interface State {
-  expanded: boolean;
-}
+export default function JSONNestedNode(props: Props) {
+  const {
+    getItemString,
+    nodeTypeIndicator,
+    nodeType,
+    data = [],
+    hideRoot,
+    createItemString,
+    styling,
+    collectionLimit,
+    keyPath,
+    labelRenderer,
+    expandable = true,
+    isCircular,
+    level = 0,
+    shouldExpandNode,
+  } = props;
 
-function getStateFromProps(props: Props) {
-  // calculate individual node expansion if necessary
-  const expanded = !props.isCircular
-    ? props.shouldExpandNode(props.keyPath, props.data, props.level)
-    : false;
-  return {
-    expanded,
-  };
-}
+  const [expanded, setExpanded] = useState<boolean>(() => {
+    return !isCircular ? shouldExpandNode(keyPath, data, level) : false;
+  });
 
-export default class JSONNestedNode extends React.Component<Props, State> {
-  static propTypes = {
-    getItemString: PropTypes.func.isRequired,
-    nodeTypeIndicator: PropTypes.any,
-    nodeType: PropTypes.string.isRequired,
-    data: PropTypes.any,
-    hideRoot: PropTypes.bool.isRequired,
-    createItemString: PropTypes.func.isRequired,
-    styling: PropTypes.func.isRequired,
-    collectionLimit: PropTypes.number,
-    keyPath: PropTypes.arrayOf(
-      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
-    ).isRequired,
-    labelRenderer: PropTypes.func.isRequired,
-    shouldExpandNode: PropTypes.func,
-    level: PropTypes.number.isRequired,
-    sortObjectKeys: PropTypes.oneOfType([PropTypes.func, PropTypes.bool]),
-    isCircular: PropTypes.bool,
-    expandable: PropTypes.bool,
-  };
-
-  static defaultProps = {
-    data: [],
-    circularCache: [],
-    level: 0,
-    expandable: true,
-  };
-
-  constructor(props: Props) {
-    super(props);
-    this.state = getStateFromProps(props);
-  }
-
-  UNSAFE_componentWillReceiveProps(nextProps: Props) {
-    const nextState = getStateFromProps(nextProps);
-    if (getStateFromProps(this.props).expanded !== nextState.expanded) {
-      this.setState(nextState);
+  // When certain props change, we need to re-compute whether our node should be in an expanded state
+  useEffect(() => {
+    setExpanded(() => {
+      return !isCircular ? shouldExpandNode(keyPath, data, level) : false;
+    });
+  }, [isCircular, data, keyPath, level, shouldExpandNode]);
```

Comment:
I believe this would reintroduce the undesired behavior discussed [here](https://github.com/alexkuz/react-json-tree/pull/61#issuecomment-280425149) since `shouldExpandNode` is often intended to only indicate whether on a node should be expanded on the first render. This was fixed by evaluating `shouldExpandNode` for the previous props and the current props and only updating the expanded state if they differ.

We're unfortunately stuck in an awkward situation where this component is partially controlled and it's unclear when `shouldExpandNode` should be called. We either need to preserve the previous behavior (which seems difficult if we don't want to keep around a reference to the previous props) or come up with a better solution that avoids the difficulties that arise from a prop that allows for partially controlled state. Feel free to pitch some ideas so we can discuss before putting too much working into a single solution. I'll keep thinking about it myself.